### PR TITLE
Delete ProxyTensor wrapper subclass

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83528
* __->__ #83330

I was working on https://github.com/pytorch/torchdynamo/issues/80 and my
working hypothesis for what was causing the error was that proxy tensor
was not advertising correct dispatch keys, causing AMP to operate
differently when you traced.  I could have fixed this directly by
replicating fake tensor's fix for setting dispatch keys to also apply to
proxy tensor, but I was like, "Why must I repeat myself."

This PR is the result.  It completely deletes the ProxyTensor wrapper
subclass, so that when we are tracing, the tensors flowing through the
program are the *original* real or fake tensors, depending on what the
user requested in the top-level API.  There is no more wrapping.  To
store the Proxy objects necessary for actually doing tracing, I store
the property directly on the tensors.  (Note: I never
clean up old entries from the map at the moment, this is easily fixed
by using a weak map)

Benefits of doing this:

* No more tip-toeing around no_dispatch() creation of new ProxyTensors;
  we never create new tensors (except when we call the underlying func),
  so you don't have to worry about accidentally tracing them.

* No more syncing up metadata from in place operators.  In particular
  https://github.com/pytorch/pytorch/issues/81526 is mooted

* This fixes https://github.com/pytorch/torchdynamo/issues/519 as we no longer need to teach proxy tensor to support sparse tensor.

* No more schlepping symbolic integers from the inner fake tensor to the
  outer proxy tensor.  If you can make a fake tensor with symbolic ints,
  you're done, nothing else to do.

To avoid having to rewrite all of the guts, when I get to the actual
proxy tensor handler, I first "fetch" the stored ProxyTensor data from
the weakmap via a tree_map, and then operate on the consequent data as
before.  A more optimized implementation is possible.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
@Chillee and I discussed this in chat, and @Chillee's main objection is that, to him, ProxyTensor makes sense as an intuitive concept, and so deleting it makes the implementation of proxy tensor less intuitive.

**Comment by eellison** [comment]
Do we have any thought on what the maintenance burden of supporting the correct dispatch keys, handling masked/sparse/nested etc, & mirroring in-place mutations will be compared to removing ProxyTensor ?

**Comment by ezyang** [comment]
We had more discussion and @Chillee is more on board with this PR, but we need to work out how symbolic ints get handled in the new world order

**Comment by IvanYashchuk** [comment]
I think having only dispatch mode without tensor subclasses for the purpose of tracing is more intuitive and simpler to use - no need to think about how to make it work correctly with other tensor subclasses. Thank you for this change, I hope it lands.

**Comment by ezyang** [comment]
SymInt tracing mode PR at https://github.com/pytorch/pytorch/pull/83380

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -93,10 +89,10 @@ def maybe_disable_fake_tensor_mode():
...
 
 
-def unwrap_elem(e):
-    if isinstance(e, ProxyTensor):
-        return e.elem
-    return e
+@dataclass
+class _ProxyTensor:
+    proxy: Proxy
+    constant: Optional[torch.Tensor]
```
Comment (eellison) [inline torch/fx/experimental/proxy_tensor.py:95]: scary to me memory wise to be carrying around a tensor... when will those be set ?

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -622,18 +573,7 @@ def get_isolated_graphmodule(func, args, kwargs):
...
-    # wrapped with another Tensor subclass
-    # See: https://github.com/pytorch/pytorch/pull/81764#issuecomment-1200472068
-    # TODO: Once https://github.com/pytorch/pytorch/pull/82549 is merged, we can
-    # remove this
-    assert all(
-        getattr(a, "elem", None) is None
-        for a in unwrapped_all_args
-        if isinstance(a, torch.Tensor)
-    ), "ProxyTensor is wrapped with another Tensor subclass"
+    unwrapped_all_args = all_args
```
Comment (eellison) [inline torch/fx/experimental/proxy_tensor.py:576]: delete this line and just `/s/unwrapped_all_args/all_args` ?

**Comment by ezyang** [comment]
All the TODOs have been fixed

File: torch/fx/experimental/proxy_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:96]: These only are ever set for torch.tensor(...) constructor call, and any subsequent operations that only have constants. So it is very, very limited. Factory functions don't count as constants.

File: torch/fx/experimental/proxy_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:578]: sure will fix after this ci round is done

**Comment by ezyang** [comment]
@albanD you may want to check on the changes to prim autograd

File: torch/_prims_common/wrappers.py [hunk 3]
```diff
@@ -286,7 +289,20 @@ def backward(ctx, *args):
...
+            # TODO: There is a subtle bug here: prims like copy_to
+            # return their input argument after mutating it; and custom
+            # autograd function will incorrectly turn the result into
+            # a view which will fail test_python_ref_executor tests.
+            # At the moment, we sidestep this by observing that the
+            # unit tests don't ever try to run the executor with
+            # autograd, so we don't exercise the buggy case, but if
+            # you ever want to feed autograd through this, be aware
+            # of it!  We need a way of properly implementing autograd
+            # for mutating operations in Python to do this.
```
Comment (albanD) [inline torch/_prims_common/wrappers.py:302]: You can use [mark_dirty()](https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.mark_dirty.html#torch.autograd.function.FunctionCtx.mark_dirty) to mark Tensors that are being modified inplace and all will work fine :)

File: torch/fx/experimental/proxy_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:101]: Yes, that's right. We can modify the policy to check the numel of the result and bail if it is bigger than one. This is a preexisting problem though, we should hit it in a separate PR

File: test/test_proxy_tensor.py [hunk 4]
```diff
@@ -284,11 +286,10 @@ def f2_logging_tensor(x):
...
             return torch.digamma(x)
 
-        with self.assertRaisesRegex(AssertionError, "ProxyTensor is wrapped with another Tensor subclass"):
-            traced = make_fx(f2_logging_tensor)(torch.randn(3))
-            self.assertFalse(is_any_sum(traced))
-            self.assertFalse(is_any_sigmoid(traced))  # this fails, sigmoid is traced with LoggingTensor
-            self.assertTrue(is_any_digamma(traced))
+        traced = make_fx(f2_logging_tensor)(torch.randn(3))
+        self.assertFalse(is_any_sum(traced))
+        self.assertFalse(is_any_sigmoid(traced))  # this fails, sigmoid is traced with LoggingTensor
```
Comment (Chillee) [inline test/test_proxy_tensor.py:291]: What's going on here?

File: torch/_decomp/decompositions.py [hunk 5]
```diff
@@ -1244,10 +1244,10 @@ def transpose_int(self: Tensor, dim0: int, dim1: int) -> Tensor:
     dim0, dim1 = utils.canonicalize_dims(self.dim(), (dim0, dim1))  # type: ignore[misc]
 
     if self.dim() <= 1:
-        return self
+        return self.view(self.shape)
```
Comment (Chillee) [inline torch/_decomp/decompositions.py:1247]: Let's add a note here?

File: aten/src/ATen/TensorSubclassLikeUtils.h [hunk 6]
```diff
@@ -39,16 +40,22 @@ constexpr auto kTensorSubclassLike =
     DispatchKeySet(BackendComponent::MetaBit);
 
 inline bool isTensorSubclassLike(const Tensor& tensor) {
+  if (c10::impl::dispatch_mode_enabled())
+    return true;
```
Comment (voznesenskym) [inline aten/src/ATen/TensorSubclassLikeUtils.h:44]: nit: parens please

File: torch/_decomp/decompositions.py (hunk 5 above)
Comment (voznesenskym) [inline torch/_decomp/decompositions.py:1247]: why does this become a view now?

File: aten/src/ATen/TensorSubclassLikeUtils.h (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/TensorSubclassLikeUtils.h:44]: The loss of parens is clang format enforced 😂. I wonder if this formatting setting is configurable

File: torch/_decomp/decompositions.py (hunk 5 above)
Comment (ezyang) [reply torch/_decomp/decompositions.py:1247]: This is a hack to guarantee we return a fresh tensor; returning an input in a view is gonna give you a bad time

**Review by Chillee — COMMENTED** [review]
Would be nice to provide some utilities for determining whether something is a proxytensor or not (like a print function or something like that).

File: test/test_proxy_tensor.py (hunk 4 above)
Comment (ezyang) [reply test/test_proxy_tensor.py:291]: The test previously failed because proxy tensor wrapping impeded subgraph support, but now that there is no wrapping, subclasses work fine!

**Review by Chillee — APPROVED** [review]
Took a closer read.

Relevant parts to me look good (i.e. `make_fx`).

File: torch/_decomp/decompositions.py (hunk 5 above)
Comment (voznesenskym) [reply torch/_decomp/decompositions.py:1247]: gotcha, do you want to add a comment for that?

**Comment by voznesenskym** [comment]
This makes cond work, please ship

**Comment by ezyang** [comment]
since this is green, I'm going to land and handle comments in follow up pr
